src/context.ts

```diff
@@ -1,76 +1,154 @@
+    confDir: string
+}
+
+export interface RockDebugConfig
+{
+    cwd: string;
+    args: string[],
+    orogen: RockOrogenDebugConfig
+}
+
+function exists(folders: vscode.WorkspaceFolder[], uri: string)
+{
+    return folders.find((item) => {
+        return (item.uri.fsPath == uri);
+    })
+}
 
 export class Context
 {
     private readonly _context: vscode.ExtensionContext;
     private readonly _vscode: wrappers.VSCode;
     private readonly _workspaces: autoproj.Workspaces;
+    private readonly _folderToPackageType: Map<string, packages.Type>;
+    private readonly _folderToDebuggingTarget: Map<string, debug.Target>;
+    private readonly _packageFactory: packages.PackageFactory;
+    private readonly _eventEmitter: vscode.EventEmitter<void>;
+    private readonly _bridge: async.EnvironmentBridge;
+    private _lastSelectedRoot: string;
 
     public constructor(context: vscode.ExtensionContext,
-                       wrapper: wrappers.VSCode, workspaces: autoproj.Workspaces)
+                       wrapper: wrappers.VSCode, workspaces: autoproj.Workspaces,
+                       packageFactory: packages.PackageFactory,
+                       eventEmitter: vscode.EventEmitter<void>,
+                       bridge: async.EnvironmentBridge)
     {
         this._context = context;
         this._vscode = wrapper;
         this._workspaces = workspaces;
+        this._packageFactory = packageFactory;
+        this._folderToDebuggingTarget = new Map<string, debug.Target>();
+        this._folderToPackageType = new Map<string, packages.Type>();
+        this._eventEmitter = eventEmitter;
+        this._bridge = bridge;
+    }
+
+    public setPackageType(path: string, type: packages.Type): void
+    {
+        let data = this.loadPersistedData(path);
+        data.type = type.name;
+        this.persistData(path, data);
+        this._eventEmitter.fire();
     }
 
-    public get selectedPackage(): { name:string, root:string }
+    public getPackageType(path: string): packages.Type
     {
+        let pkgType: packages.Type;
+        let data = this.loadPersistedData(path);
+
+        if (data.type)
+            pkgType = packages.Type.fromName(data.type);
+
+        return pkgType;
+    }
+
+    public setDebuggingTarget(path: string, target: debug.Target): void
+    {
+        let data = this.loadPersistedData(path);
+        data.debuggingTarget.name = target.name;
+        data.debuggingTarget.path = target.path;
+        this.persistData(path, data);
+        this._eventEmitter.fire();
+    }
+
+    public getDebuggingTarget(path: string): debug.Target
+    {
+        let data = this.loadPersistedData(path);
+        if (!data || !data.debuggingTarget ||
+            !data.debuggingTarget.name || !data.debuggingTarget.path)
+            return undefined;
+
+        return new debug.Target(data.debuggingTarget.name,
+            data.debuggingTarget.path);
+    }
+
+    public async getSelectedPackage(): Promise<packages.Package>
```

Comment:
I have the impression here that you assume that every editor/path is a package. Does that really make any sense at all ?

Reply:
I assume that every folder added to the vscode's workspace are either a configuration (autoproj/ and its childs) or a package. Then, getWorkspaceFolder(resource) gives me the parent folder (which I assume is a package). Thoughts?

Reply:
> I assume that every folder added to the vscode's workspace are either a configuration (autoproj/ and its childs) or a package. 

I don't think it makes sense. People may have to add other packages outside of their workspace. I think it's taken care of in the refactoring I'm doing (through InvalidPackage).

Reply:
Current implementation allows packages from outside of an workspace. They are [ForeignPackages](https://github.com/g-arjones/vscode-rock/blob/autoproj/src/packages.ts#L362). Anyway, let's wait for your refactoring...